src/olympia/discovery/tests/test_admin.py

```diff
@@ -546,6 +551,164 @@ def test_query_count(self):
+        user = user_factory()
+        self.grant_permission(user, 'Admin:Tools')
+        self.grant_permission(user, 'Discovery:Edit')
+        self.client.login(email=user.email)
+        response = self.client.get(self.list_url, follow=True)
+        assert response.status_code == 200
+        assert 'föo.jpg' in response.content.decode('utf-8')
+
+    def test_can_edit_with_discovery_edit_permission(self):
+        item = PrimaryHeroImage.objects.create(custom_image='föo.jpg')
+        self.detail_url = reverse(
+            'admin:discovery_primaryheroimageupload_change', args=(item.pk,)
+        )
+        user = user_factory()
+        self.grant_permission(user, 'Admin:Tools')
+        self.grant_permission(user, 'Discovery:Edit')
+        self.client.login(email=user.email)
+        response = self.client.get(self.detail_url, follow=True)
+        assert response.status_code == 200
+        content = response.content.decode('utf-8')
+        assert u'föo.jpg' in content
+
+        response = self.client.post(
+            self.detail_url,
+            dict(self._get_heroform(str(item.id)), **{
+                'custom_image': 'föo.jpg',
+            }),
+            follow=True)
+        assert response.status_code == 200
+        # item.reload()
+        assert PrimaryHeroImage.objects.count() == 1
+        assert item.custom_image == 'föo.jpg'
+
+    def test_can_delete_with_discovery_edit_permission(self):
+        item = PrimaryHeroImage.objects.create()
+        delete_url = reverse(
+            'admin:discovery_primaryheroimageupload_delete', args=(item.pk,)
+        )
+        user = user_factory()
+        self.grant_permission(user, 'Admin:Tools')
+        self.grant_permission(user, 'Discovery:Edit')
+        self.client.login(email=user.email)
+        # Can access delete confirmation page.
+        response = self.client.get(delete_url, follow=True)
+        assert response.status_code == 200
+        assert PrimaryHeroImage.objects.filter(pk=item.pk).exists()
+
+        # And can actually delete.
+        response = self.client.post(
+            delete_url,
+            dict(self._get_heroform(str(item.id)), post='yes'),
+            follow=True)
+        assert response.status_code == 200
+        assert not PrimaryHeroImage.objects.filter(pk=item.pk).exists()
+
+    def test_can_add_with_discovery_edit_permission(self):
+        add_url = reverse('admin:discovery_primaryheroimageupload_add')
+        user = user_factory()
+        self.grant_permission(user, 'Admin:Tools')
+        self.grant_permission(user, 'Discovery:Edit')
+        self.client.login(email=user.email)
+        response = self.client.get(add_url, follow=True)
+        assert response.status_code == 200
+        assert PrimaryHeroImage.objects.count() == 0
+        response = self.client.post(
+            add_url,
+            dict(self._get_heroform(''), **{
+                'custom_image': '',
```

Comment:
>  I feel like there should be a sample filename for custom_image, but I kept getting an error if I included one

What's the error?  Is it because you aren't actually including the file contents in the POST?  I suspect that may be necessary. 

I'm not sure what the request data structure would look like (if you select an image file in the form and then submit the form to upload it, devtools should show you) but it's probably something like `open(<path to image somewhere under tests />, 'rb')` and then `'upload': data` in the dict.

The `<path to image somewhere under tests/ >` you can probably use `olympia.amo.tests.test_helpers.get_image_path` for now.

Reply:
I think you were right about it being that I did not include the file contents. The error is gone now. :)

Could I use `olympia.amo.tests.test_helpers.get_uploaded_file`?